core/src/main/java/org/elasticsearch/threadpool/Scheduler.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.threadpool;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.logging.log4j.util.Supplier;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.EsAbortPolicy;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Scheduler that allows to schedule one-shot and periodic commands.
+ */
+public class Scheduler extends AbstractComponent implements Closeable {
```

Comment:
I was thinking of an interface that is part of the bulk processor like:
```
interface Scheduler {
  ScheduledFuture<?> schedule(TimeValue delay, String executor, Runnable command);
  Cancellable scheduleWithFixedDelay(Runnable command, TimeValue interval, String executor);
  void shutdown();
}
```

That way core doesn't depend on bulk processor stuff, bulk processor stuff depends on core. Or at least can be adapted into core.

And when we want to work with a ThreadPool we could implement it like:
```
  new Scheduler() {
    ScheduledFuture<?> schedule(TimeValue delay, String executor, Runnable command) {
      return threadPool.schedule(delay, executor, command);
    }

    Cancellable scheduleWithFixedDelay(Runnable command, TimeValue interval) {
      ThreadPool.Cancellable c = threadPool.scheduleWithFixedDelay(threadPool.preserveContext(command), interval, SAME);
      return new Cancellable() {
        void cancel() {c.cancel();}
        boolean isCancelled() {return c.isCancelled();}
    }

    void shutdown() {
      // Do nothing because we don't want to shut down the shared ThreadPool.
    }
  }
```

Is that crazy? It feels like that lets users plug it into whatever scheduling solution they want.

Reply:
I would prefer if users didn't have to do anything. I did try the interface idea, but I quickly realized that I need the same code in both scheduler impls (the one used by ThreadPool and the one that BulkProcessor uses).

How would you implement scheduleWithFixedDelay in the bulk processor impl? This question led me to the current impl. Some of the reasons are in the description of this PR. From then on, I had to add some more methods to the base class, but in hindsight it kinda makes sense to have the scheduler bits extracted out and shared, I thought.

Reply:
Something like:

```
class ThreadPoolExecutorScheduler implements Scheduler {
  private final ScheduledThreadPoolExecutor exec;

  public ScheduledFuture<?> schedule(TimeValue delay, String executor, Runnable command) {
      return exec.schedule(delay, executor, command);
    }

    public Cancellable scheduleWithFixedDelay(Runnable command, TimeValue interval) {
      ScheduledFuture<?> c = threadPool.scheduleWithFixedDelay(command, interval, SAME);
      return new Cancellable() {
        void cancel() {c.cancel();}
        boolean isCancelled() {return c.isCancelled();}
    }

    public void shutdown() {
      exec.shutdown();
    }
  }
```

I really don't want stuff like `Settings` and `AbstractComponent` to make their way too far into the high level rest client if we can help it. I mean, `ThreadPool` is even worse because it is all of those things, so the change as you have it is an improvement.

I think the user doesn't really have to do anything. If the BulkProcessor by default builds it own scheduler building a `ScheduledThreadPoolExecutor` on the fly then great. If they call `builder.withScheduler(ScheduledThreadPoolExecutor)` then we can build a scheduler for them. And if they want ultimate control they can call `builder.withScheduler(Scheduler)`. And we can have a `builder.withSchedulerFromThreadPool(threadPool)` for transport client users.

Reply:
Er, I was using bad names in the comment above. Edited.

Reply:
Talked with @javanna - the trouble with what I propose is that `ScheduledThreadPoolExecutor` doesn't react the same way to failures as `ThreadPool`. Especially with regards to `scheduleWithFixedDelay` style stuff. In `ThreadPool` we retry even if it fails. `ScheduledThreadPoolExecutor` doesn't. We think.

Reply:
I think you meant to call `exec.scheduldWithFixedDelay` from `scheduleWithFixedDelay`. The problem is that "If any execution of the task encounters an exception, subsequent executions are suppressed." In our own impl (currently in `ThreadPool`) a failure won't affect subsequent runs, which will still happen. That is the main reason why I didn't go for an interface, and move some more methods to the base impl, which can be shared between ES core and the high-level client. I do see how AbstractComponent is annoying, together with Settings, but that is a problem that we may want to fix in ES core directly and probably just a small downside in the context of this change?